### PR TITLE
[SPARK-31916][SQL] StringConcat can lead to StringIndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by dilipbiswal**
### What changes were proposed in this pull request?
A minor fix to fix the append method of StringConcat to cap the length at MAX_ROUNDED_ARRAY_LENGTH to make sure it does not overflow and cause StringIndexOutOfBoundsException

Thanks to **Jeffrey Stokes** for reporting the issue and explaining the underlying problem in detail in the JIRA.

### Why are the changes needed?
This fixes StringIndexOutOfBoundsException on an overflow.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Added a test in StringsUtilSuite.

### CODE REVIEW

**Comment by dilipbiswal** [comment]
cc @maropu

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala [hunk 1]
```diff
@@ -98,4 +98,12 @@ class StringUtilsSuite extends SparkFunSuite {
     assert(checkLimit("1234567"))
     assert(checkLimit("1234567890"))
   }
+
+  test("StringConcat doesn't overflow on many inputs") {
```
Comment (kiszk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:102]: nit: could you please add the prefix: "SPARK-31916: " ?

Comment (dilipbiswal) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:102]: @kiszk Thanks.. will do.

**Review by MaxGekk — COMMENTED** [review]
After your changes, the error message https://github.com/apache/spark/pull/28750/files#diff-46beff422d36d30d696d137ae1cfcf2eR151 will not show the actual size.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 2]
```diff
@@ -122,8 +122,8 @@ object StringUtils extends Logging {
           val available = maxLength - length
           val stringToAppend = if (available >= sLen) s else s.substring(0, available)
           strings.append(stringToAppend)
+          length += sLen
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:125]: nit: Can we move line 120 into `{...}` to make the scope sure?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 3]
```diff
@@ -122,8 +122,8 @@ object StringUtils extends Logging {
...
         }
-        length += sLen
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: How about:
```scala
        length = Math.min(length.toLong + sLen, ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH).toInt
```

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: @MaxGekk Yeah.. Sure. Just wondering we absolutely need to change the length in case we are not doing an append. The error message you are referring to, is this captured in some test ? I ran catalyst and sql tests and didn't see any failure.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: > The error message you are referring to, is this captured in some test ?

Looks like not. Could you add a test for that?

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: @MaxGekk Sure.. Will add a test. Can you please point me to the error message please ? The link you have, when i click .. its not taking me to the relevant code ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:125]: @kiszk If i incorporate @MaxGekk's suggestion, then i will need to keep 120 where it is :-) as we need it outside the if block.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: In the child class, here https://github.com/apache/spark/blob/d9b30694122f8716d3acb448638ef1e2b96ebc7a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala#L156
https://github.com/apache/spark/blob/d9b30694122f8716d3acb448638ef1e2b96ebc7a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala#L151

**Review by MaxGekk — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala (hunk 3 above)
Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: Nitpicking, but:
Why not only update length inside the `if (!atLimit)` block as originally shown?
And only add `stringToAppend.length()` there, if the point of `Math.min()` is to avoid overflow if adding a massive string where only part was added.

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: @srowen I think, the original intent of keeping this length outside the if block (which i missed and @MaxGekk pointed out) is to keep the true length of the input. So when we are producing a SQL plan, we want to tell users how much was that the size of the explain string and how much we truncated based on max length specification. So caller could call append method even after we have gone past the max. 
cc @MaxGekk

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: Oh I see. If overflow is an issue, then make it a long? Capping it at this constant would then seem to potentially underreport the actual appended total.
Maybe that creates other problems, not sure.

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: @srowen You are right that, we may under report when the length becomes more than MAX_ROUNDED_ARRAY_LENGTH. I did consider changing the message to cover the case to say something like "greater than or equal to N". But thought against it as we are adding code to cover may be a extremely corner case. Please let me know if you have any suggestion here.

About changing to long, even if we did it, we can in theory hit the long limit, right ?

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: I'm not sure you can overflow a long - if 2^31-1 strings (max length of the array tracking them) are added, of length 2^31-1, you still don't overflow.
It's probably not worth debating so this is fine but I think a simpler change was also fine.

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: Thank you @srowen.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala [hunk 4]
```diff
@@ -98,4 +98,12 @@ class StringUtilsSuite extends SparkFunSuite {
     assert(checkLimit("1234567"))
     assert(checkLimit("1234567890"))
   }
+
+  test("SPARK-31916: StringConcat doesn't overflow on many inputs") {
+    val concat = new StringConcat(maxLength = 100)
+    0.to(Integer.MAX_VALUE).foreach { _ =>
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:104]: Could you reduce # of loops in this test? I think its a bit expensive.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 5]
```diff
@@ -3503,6 +3504,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     checkAnswer(sql("select CAST(-32768 as short) DIV CAST (-1 as short)"),
       Seq(Row(Short.MinValue.toLong * -1)))
   }
+
+
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3508]: nit: remove the single blank.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 6]
```diff
@@ -3503,6 +3504,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+  test("SPARK-31916: verify that PlanStringConcat's output shows the actual length of the plan") {
+    withSQLConf("spark.sql.maxPlanStringLength" -> "0") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3510]: plz use `SQLConf.MAX_PLAN_STRING_LENGTH.key` instead.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 7]
```diff
@@ -3503,6 +3504,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+      val concat = new PlanStringConcat()
+      0.to(3).foreach { i =>
+        concat.append(s"plan fragment $i")
+      }
+      assert(concat.toString === "Truncated plan of 60 characters")
+    }
+
+    withSQLConf("spark.sql.maxPlanStringLength" -> "60") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3518]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 8]
```diff
@@ -3503,6 +3504,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     checkAnswer(sql("select CAST(-32768 as short) DIV CAST (-1 as short)"),
       Seq(Row(Short.MinValue.toLong * -1)))
   }
+
+
+  test("SPARK-31916: verify that PlanStringConcat's output shows the actual length of the plan") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3509]: Since `PlanStringConcat's` is in the catalyst package, could you move this tests there?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:126]: Ah, could you leave some comments about the intent here? I think its a bit hard to tell it at a glance.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 9]
```diff
@@ -29,6 +29,7 @@ import org.apache.spark.sql.catalyst.expressions.aggregate.{Complete, Partial}
 import org.apache.spark.sql.catalyst.optimizer.{ConvertToLocalRelation, NestedColumnAliasingSuite}
 import org.apache.spark.sql.catalyst.plans.logical.Project
 import org.apache.spark.sql.catalyst.util.StringUtils
+import org.apache.spark.sql.catalyst.util.StringUtils.PlanStringConcat
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:32]: unnecessary change?

**Review by viirya — COMMENTED** [review]
Please also update the PR description as looks like it is still for original fix.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
@dilipbiswal Have you checked my last comment? https://github.com/apache/spark/pull/28750#discussion_r437061918 The PR itself looks okay.

**Review by viirya — COMMENTED** [review]
Looks okay although I think making it a long might be also good and simpler.

**Comment by dilipbiswal** [comment]
@maropu 
> Have you checked my last comment? #28750 (comment) The PR itself looks okay.

Sorry i missed that. I have added the comment now.

@viirya 
We could make it a long. The only thing is we may still overflow but it will take perhaps long time to hit it. I can repro the StringIndexOutOfBoundsException after changing the length to long. I made a minor tweak to change the append method to fake the input's length to be Int.MaxValue and adjust the test to increase the loop count.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 10]
```diff
@@ -123,7 +123,11 @@ object StringUtils extends Logging {
           val stringToAppend = if (available >= sLen) s else s.substring(0, available)
           strings.append(stringToAppend)
         }
-        length += sLen
+
+        // Cap the length at  ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH. Otherwise, we
+        // will overflow length causing StringIndexOutOfBoundsException in the substring call
+        // above.
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:129]: @dilipbiswal Sorry and my comment looked ambiguous. I just wanted to leave a note about your comment above;
>  I think, the original intent of keeping this length outside the if block (which i missed and @MaxGekk pointed out) is to keep the true length of the input. So when we are producing a SQL plan, we want to tell users how much was that the size of the explain string and how much we truncated based on max length specification. So caller could call append method even after we have gone past the max.

I couldn't tell why the count-up is placed outside `the if (!atLimit) block`.

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:129]: @maropu Ah .. so you wanted to explain the above comment better in the PR and not in the code ? Right ?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:129]: The comment in the code side looks okay and how about this?
```
        // Keeps the total length of appended strings. Note that we need to cap the length at
        // `ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH`; otherwise, we will overflow
        // length causing StringIndexOutOfBoundsException in the substring call above.
        length = Math.min(length.toLong + sLen, ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH).toInt
```

**Comment by maropu** [comment]
retest this please

**Review by maropu — APPROVED** [review]

**Comment by dilipbiswal** [comment]
retest this please

**Review by srowen — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master/branch-3.0.
@dilipbiswal Could you make a backport PR to branch-2.4?

**Comment by dilipbiswal** [comment]
@maropu Sure. Thanks a lot for your comments on the pr.

**Comment by maropu** [comment]
Thanks!

**Comment by dilipbiswal** [comment]
Hi @maropu 
Just checked the 2.4 codebase. We don't have this Class in 2.4. Looks like we directed appended to StringBuilder.  This class was added as part of [SPARK-26504](https://issues.apache.org/jira/browse/SPARK-26504)

Please let me know.

**Comment by maropu** [comment]
Ah, okay. We don't need the backport. Thanks for the check, anyway!
